### PR TITLE
opt: do not simplify window func with RANGE mode and offset

### LINKED ISSUE #53442: opt: unexpected behavior in an edge case with RANGE mode of window framing
Found by sqlsmith in #48268.

The following results in a panic
```
CREATE TABLE table1 (col1 INT8 PRIMARY KEY);

CREATE TABLE table2 (col2 INT2 PRIMARY KEY);

SELECT
  max(col2::INT8) OVER (
    PARTITION BY col2 ORDER BY col2 RANGE 1 PRECEDING
  )
FROM
  table2 NATURAL JOIN table1
WHERE
  false;
```
This is a pathological case with RANGE mode of framing in which ORDER BY clause (which must contain exactly one column) contains the same column as the one mentioned in PARTITION BY clause, and such case was supposed to be fixed by #44666. Interestingly, if I remove the cast within `max`, then the plan is different and the panic doesn't occur.

The panic occurs because at the execution time that single ORDER BY column has `-1` as its index which is populated incorrectly in https://github.com/cockroachdb/cockroach/blob/96e89e091a278b9938d35060b9c1b3583125af49/pkg/sql/opt/exec/execbuilder/relational.go#L2190-L2193
This occurs, I think, because the optimizer thinks that `col2` column is not needed in the output, so `w.RangeOffsetColumn == 1` gets mapped to `-1`.

cc @RaduBerinde

### PR DESCRIPTION
**Pull request by mgartner**
This commit updates the SimplifyWindowOrdering normalization rule so
that it does not apply to window functions that have a RANGE mode with a
specific frame offset boundary of PRECEDING or FOLLOWING. The execution
engine panics if an ordering column is not provided in this case, even
if the ordering is a constant ordering.

This provides a more robust solution to the original issue that #44666
attempted to fix. That solution did not cover all edge cases because it
did not consider that the stored RangeOffsetColumn could be pruned if
there were no other references to it. This case is rare and it is often
hidden by other optimizations that eliminate a window function with a
cardinality of zero. But it is possible; see the added logic tests for
examples.

This commit reverts some of the changes from #44666 because they are no
longer necessary.

Fixes #53442

Release justification: This is a high-priority fix for a bug in WINDOW
functions that causes panics.

Release note (bug fix): A bug was fixed that caused a crash when using a
RANGE mode window function with an offset in some cases, e.g, OVER
(PARTITION BY b ORDER BY a RANGE 1 PRECEDING).

### CODE REVIEW

**Review by yuzefovich — APPROVED** [review]
I have one comment, but otherwise :lgtm:

> Reviewed 10 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53717#-:-MG5kPvk1r0mgeL9fEvu:b2kx0us)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/opt/memo/expr.go, line 525 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53717#-MG5jto_8AuJk0wI77nL:-MG5jto_8AuJk0wI77nM:b-21euq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4628dad36a769a220de3d7a59cb36debc6b0b4b3/pkg/sql/opt/memo/expr.go#L525)):*
> ```Go
> // HasOffset returns true if the WindowFrame contains a distinct offset.
> func (f *WindowFrame) HasOffset() bool {
> 	return f.StartBoundType == tree.OffsetPreceding || f.EndBoundType == tree.OffsetFollowing
> ```

Both start and end bounds can be of both `PRECEDING` and `FOLLOWING` types (e.g. `BETWEEN UNBOUNDED PRECEDING AND 1 PRECEDING` and `BETWEEN 1 FOLLOWING AND 3 FOLLOWING`), so we need two more cases.
___
*[pkg/sql/opt/norm/window_funcs.go, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53717#-MG5kDAh1vhK3nl9weve:-MG5kDAh1vhK3nl9wevf:b-gkk6gd) ([raw file](https://github.com/cockroachdb/cockroach/blob/4628dad36a769a220de3d7a59cb36debc6b0b4b3/pkg/sql/opt/norm/window_funcs.go#L154)):*
> ```Go
> 
> // HasRangeFrameWithOffset returns true if w contains a WindowsItem Frame that
> // has a mode of RANGE and has a distinct offset (the start bound is PRECEDING
> ```

nit: this comment will need an adjustment according to fix to `HasOffset`.

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53717)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde and @yuzefovich)

___
*[pkg/sql/opt/memo/expr.go, line 525 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53717#-MG5jto_8AuJk0wI77nL:-MG5prRo8g0n4xUppAZi:b11ocsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4628dad36a769a220de3d7a59cb36debc6b0b4b3/pkg/sql/opt/memo/expr.go#L525)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Both start and end bounds can be of both `PRECEDING` and `FOLLOWING` types (e.g. `BETWEEN UNBOUNDED PRECEDING AND 1 PRECEDING` and `BETWEEN 1 FOLLOWING AND 3 FOLLOWING`), so we need two more cases.
</blockquote></details>

Doh! I was looking at the code and it seemed to allow both PRECEDING and FOLLOWING for both start and end bounds, but wasn't sure what SQL concoction made use of it. Updated.
___
*[pkg/sql/opt/norm/window_funcs.go, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53717#-MG5kDAh1vhK3nl9weve:-MG5q6XYANrkM-d6I1ym:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4628dad36a769a220de3d7a59cb36debc6b0b4b3/pkg/sql/opt/norm/window_funcs.go#L154)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: this comment will need an adjustment according to fix to `HasOffset`.
</blockquote></details>

Done.

**Review by yuzefovich — APPROVED** [review]
> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53717)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

**Review by yuzefovich — COMMENTED** [review]
nit: maybe it's worth calling this out in the release note (also, I think this commit should be backported).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53717#-:-MG5txA34rvnRNntR9p4:b-wl2t3j)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

**Review by mgartner — COMMENTED** [review]
> nit: maybe it's worth calling this out in the release note (also, I think this commit should be backported).

Done.

I can backport once this is approved. I think this would apply to 19.2 and 20.1, not sure about 19.1.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53717#-:-MGA37yw4uSQQWlEw08G:b6ld3cl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde and @yuzefovich)

**Comment by yuzefovich** [comment]
I wouldn't worry about 19.1 because it is reaching its end of life soon.

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53717#-:-MGAgJxJ39AGqZEZjr35:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @yuzefovich)

___
*[pkg/sql/opt_exec_factory.go, line 1001 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/53717#-MGAg2KK-yay3X4ozdRH:-MGAg2KK-yay3X4ozdRI:b-q0w076) ([raw file](https://github.com/cockroachdb/cockroach/blob/664a984b95cf0582d7fbf6b0377370136a89af46/pkg/sql/opt_exec_factory.go#L1001)):*
> ```Go
> 			frame := p.funcs[i].frame
> 			if frame.Mode == tree.RANGE && frame.Bounds.HasOffset() {
> 				// We have an empty ordering, but RANGE mode when at least one bound
> ```

This case is no longer possible, right? If so, I'd leave the check and add an assertion error to future-proof
___
*[pkg/sql/opt/norm/window_funcs.go, line 154 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/53717#-MGAg9Qm68U2f_JeAMQU:-MGAg9Qm68U2f_JeAMQV:b-m7iqzs) ([raw file](https://github.com/cockroachdb/cockroach/blob/d98e99bf47697815e37c828c7676fc1d3483df32/pkg/sql/opt/norm/window_funcs.go#L154)):*
> ```Go
> 
> // HasRangeFrameWithOffset returns true if w contains a WindowsItem Frame that
> // has a mode of RANGE and has a distinct offset, such as OffsetPreceding or
> ```

what's a "distinct offset"? Maybe "specific offset"?

**Review by RaduBerinde — COMMENTED** [review]
In 19.1, the optimizer did not support window functions at all.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53717#-:-MGAgSoM-CDIh3vwN91B:b-xmdtsd)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @yuzefovich)

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53717)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)

___
*[pkg/sql/opt_exec_factory.go, line 1001 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/53717#-MGAg2KK-yay3X4ozdRH:-MGApcNrEtZrQC_yjMjJ:bcvjgzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/664a984b95cf0582d7fbf6b0377370136a89af46/pkg/sql/opt_exec_factory.go#L1001)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This case is no longer possible, right? If so, I'd leave the check and add an assertion error to future-proof
</blockquote></details>

Good idea. Done.
___
*[pkg/sql/opt/norm/window_funcs.go, line 154 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/53717#-MGAg9Qm68U2f_JeAMQU:-MGApwyc4IBJ5Chx14w3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d98e99bf47697815e37c828c7676fc1d3483df32/pkg/sql/opt/norm/window_funcs.go#L154)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

what's a "distinct offset"? Maybe "specific offset"?
</blockquote></details>

Done.

**Pull request merged**
